(add) dataset-ML

ysooj opened the pull request:
머신러닝
---------
머신러닝 노션 url에서 텍스트를 읽어오고

읽어온 텍스트를 전처리했습니다.

전처리까지 완료한 텍스트를 로컬 폴더에 저장했습니다.

ysooj committed:
(add, 머신러닝 데이터셋 불러오기)
토글 부분은 성진님 코드 참고했습니다.

ysooj committed:
전처리 작업 추가
불필요한 문구(예: [스파르타코딩클럽], 저작권 문구, 모든 토글 닫기 단축키) 제거
불필요한 공백, 특수 문자, 줄바꿈 제거

ysooj committed:
전처리 후 txt 파일로 로컬 폴더(cleaned_texts)에 저장하기
폴더가 없으면 만들고, txt 파일 이름을 설정해준다.

ysooj committed:
(add) 대화를 txt 파일로 저장하고, feedback에 불러오기
(fix) feedback_prompt, invoke 호출 방식, input_data 수정

ysooj reviewed (commented):
대화 내용을 txt 파일로 저장하고, txt 파일을 불러와서 피드백에 적용하는 코드를 추가했습니다.

##### 1.대화 내용 저장
사용자가 입력한 퀴즈(quiz)와 답변(answer), 그리고 AI가 제공한 피드백(feedback)을 파일에 기록합니다.

- previous_conversation라는 폴더가 존재하지 않으면 생성합니다.
- 대화 내용을 저장할 파일 이름은 현재 시각(timestamp)을 기준으로 생성됩니다.
- 각 대화가 끝날 때마다 파일을 열고, 퀴즈, 사용자 답변, 피드백을 기록합니다.
- 파일 열 때는 "a" 모드(append)를 사용하여 기존 내용에 새 대화를 추가합니다.

```python
# 대화 내용 저장(파일에 기록)
with open(file_path, "a", encoding="utf-8") as f:
    f.write(f"Quiz: {quiz}\n")               # 생성된 퀴즈 저장
    f.write(f"User Answer: {user_answer}\n") # 사용자 답변 저장
    f.write(f"Feedback: {feedback}\n")       # AI 피드백 저장
    f.write("-" * 50 + "\n")                 # 구분선 추가
```
--------------------
##### 2.대화 내용 불러오기
이전 대화 기록을 파일에서 읽어와 현재 대화에 활용합니다.

- 파일이 존재하면 내용을 읽어옵니다.
- 파일이 없으면 빈 문자열("")을 기본값으로 설정합니다.
- "r" 모드(read)로 파일을 열어 모든 내용을 불러옵니다.

```python
if os.path.exists(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        previous_conversation = f.read()  # 이전 대화 기록 불러오기
else:
    previous_conversation = ""           # 파일이 없으면 빈 문자열로 초기화
```
---------------------
---------------------
### 문제점
#### 문제 1
feedback_prompt에 불러온 txt 파일을 적용하는 과정에서 오류가 발생했습니다.   
    
오류 : KeyError: 'Input to ChatPromptTemplate is missing variables {"\'refusal\'", "\'input_tokens\'", "\'token_usage\'"}.  Expected: ["\'input_tokens\'", "\'refusal\'", "\'token_usage\'"] Received: [\'quiz\', \'answer\']\nNote: if you intended {\'refusal\'} to be part of the string and not a variable, please escape it with double curly braces like: \'{{\'refusal\'}}\'.\nFor troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/INVALID_PROMPT_INPUT '   
   
ChatPromptTemplate이 기대하는 입력 변수와 실제로 제공된 입력 간의 불일치때문에 발생한 오류라고 합니다.   
input_data에 필요한 변수들을 넣어주고, 피드백 체인을 호출할 때 input_data를 넣어주었습니다.
```python
# 필요한 변수를 기본값으로 추가
    input_data = {
        "quiz": quiz,
        "answer": user_answer,
        "previous_conversation": previous_conversation,
        "refusal": "",  # 기본값 설정
    }

    # 피드백 체인 호출
    feedback_chain = feedback_prompt | llm
    feedback = feedback_chain.invoke(input_data)
```  
오류 코드에서는 'input_tokens', 'token_usage', 'refusal'가 필요하다고 나와있었습니다.   
'input_tokens'와 'token_usage'는 추가하지 않아도 코드가 잘 실행돼서 따로 추가하지 않았습니다.   
그러나 'refusal'를 추가하지 않으면 계속 KeyError가 발생해서 refusal="None" 변수를 추가했습니다.
---------------------------
#### 문제 2
변수를 추가한 후에도 TypeError, ValueError가 발생했습니다.   
최종적으로 feedback_prompt를 수정하고, invoke 호출 방식을 변경하고, input_data를 수정했습니다.   
```python
# 3. 사용자 답변에 대한 피드백 생성
    feedback_prompt = ChatPromptTemplate.from_messages([
        ("system", f"""
    AI 강사로서 다음 퀴즈의 정답 여부를 확인하고 피드백을 제공하세요.
    퀴즈: {{quiz}}
    답변: {{answer}}
    대화 기록: {{previous_conversation}}
    거절 사유: {{refusal}}
    """)
    ])

    # 피드백 생성 - 키워드 인수로 전달
    feedback_data = feedback_prompt.format(
        quiz=quiz,
        answer=user_answer,
        previous_conversation=previous_conversation,
        refusal="None"
    )

    # 피드백 체인 호출
    feedback = llm.invoke(feedback_data)   # LLM을 직접 호출하여 피드백 생성
```

KooSuYeon commented:
@ysooj 
감사합니다! 트러블 슈팅까지 남겨주셔서 이해하기 편했어요! :)


feedback GPT feedback GPT에게는 previous_conversation 보다는 이전에 GPT가 생성했던 질문인지만 확인하면 될 것 같아서요..!
기존에 previous_conversation을 불러왔던 부분을 question만 불러오는 것으로 수정해주시면 갑사하겠습니다!

```
quiz_list = []

while True: 
    quiz = rag_chain.invoke("퀴즈를 시작하세요.")
    quiz_list.append(quiz)
    # 3. 사용자 답변에 대한 피드백 생성
    feedback_prompt = ChatPromptTemplate.from_messages([
        ("system", f"""
        AI 강사로서 다음 퀴즈의 정답 여부를 확인하고 피드백을 제공하세요.
        피드백은 아래와 같은 형식이어야 합니다:
        - 정답 여부: "N번" 또는 "예/아니오"
        - 추가 설명: (정답과 관련된 추가 정보를 제공하세요)
         
        퀴즈: {quiz}
        사용자의 답변: {user_answer}
        피드백:
        """)
    ])
    feedback_chain = feedback_prompt | llm
    feedback = feedback_chain.invoke({"quiz": quiz, "answer": user_answer})
    print("Feedback:")
    pprint(feedback.content)

```

ysooj committed:
fix/quiz의 질문 부분만 추출해서 feedback_prompt에 전달하기
- 정규식을 사용해 특정 패턴만 추출
- feedback_prompt의 '대화 기록'부분과 모델에 전달하는 데이터의 previous_conversation을 quiz_onlys로 수정

ysooj reviewed (commented):
@KooSuYeon 
previous_conversation에서 quiz의 질문 부분만 추출해서 feedback_prompt에 전달하도록 코드 수정했습니다!

ysooj committed:
fix/리뷰에 제공한 코드를 바탕으로 코드 수정
- quiz_list 생성
- quiz의 질문 부분 내용을 추출해오는 곳 수정

ysooj reviewed (commented):
@KooSuYeon 
제공해주신 코드를 바탕으로 다시 수정했습니다!

KooSuYeon commented:
@ysooj 
해당 브랜치는 main 브랜치에 merge되었기 떄문에 시간 나실 때 close 해주시면 감사하겠습니다!!

ysooj commented:
@KooSuYeon 
브랜치를 삭제하라는 말씀이신거죠?

KooSuYeon commented:
@ysooj 
넵! 

보통은 close할 때 반영사항이 main브랜치에 반영이 되어서 닫는다 or 필요없는 내용이라 닫는다 하고

아래 사진에서 Close with comment 하고 닫습니다! (최대한 공유하면서 PR 관리하기 위해서요!)
그 후에 delete branch 해주시면 됩니다!

![Screenshot 2024-11-27 at 12 58 45 PM](https://github.com/user-attachments/assets/916e2924-ef0f-49d8-9f9b-643a57dfcbc6)


@kyso8575 @sj5black @HSgoon 참고 부탁드릴게요!

ysooj commented:
main브랜치에 반영이 되어서 닫는다
